pkg/operator/staticpod/controller/missingstaticpod/missing_static_pod_controller.go

```diff
@@ -0,0 +1,260 @@
+		installerPodRevision, err := internal.GetRevisionOfPod(latestInstallerPodOnNode)
+		if err != nil {
+			// we expect every installer pod to have a installerPodRevision in its name, unexpected error here
+			return fmt.Errorf("failed to get installerPodRevision for installer pod %q - %w", latestInstallerPodOnNode.Name, err)
+		}
+
+		finishedAt, ok := installerPodFinishedAt(latestInstallerPodOnNode)
+		if !ok {
+			// either it's in the process of running, or it ran into an error
+			continue
+		}
+
+		gracePeriod, err := c.getStaticPodTerminationGracePeriodSecondsForRevision(installerPodRevision)
+		if err != nil {
+			return err
+		}
+		threshold := gracePeriod + 120*time.Second
+
+		staticPodRevisionOnThisNode := getStaticPodCurrentRevision(node, originalOperatorStatus)
+		if staticPodRevisionOnThisNode == 0 {
+			// skip the bootstraping phase
+			continue
+		}
+
+		if time.Since(finishedAt) > threshold &&
+			staticPodRevisionOnThisNode < installerPodRevision {
+			// if we are here:
+			//  a: the latest installer pod successfully completed at finishedAt
+			//  b: it has been more than 'terminationGracePeriodSeconds' + 120s since the
+			//     installer pod has completed at finishedAt
+			//  c. the static pod is not at correct installerPodRevision yet
+			// then all of the above conditions are true, so we should produce an event
+
+			// only emit an event if for a particular node:
+			// - this is the first time we see a failure
+			// - this is a failure for new revision
+			// - the previously reported event was at least 30 min ago
+			lastEventEmission, found := c.lastEventEmissionPerNode[node]
+			if !found || installerPodRevision != lastEventEmission.revision || time.Since(lastEventEmission.timestamp) > 30*time.Minute {
+				syncCtx.Recorder().Eventf("MissingStaticPod", "static pod lifecycle failure - static pod: %q in namespace: %q for revision: %d on node: %q didn't show up, waited: %v",
+					c.operandName, c.targetNamespace, installerPodRevision, node, threshold)
+
+				lastEventEmission.revision = installerPodRevision
+				lastEventEmission.timestamp = time.Now()
+				c.lastEventEmissionPerNode[node] = lastEventEmission
+			}
+
+			errors = append(errors, fmt.Sprintf("static pod lifecycle failure - static pod: %q in namespace: %q for revision: %d on node: %q didn't show up, waited: %v",
+				c.operandName, c.targetNamespace, installerPodRevision, node, threshold))
+		}
+	}
+
+	if len(errors) > 0 {
+		return fmt.Errorf(strings.Join(errors, "\n"))
+	}
+
+	return nil
+}
+
+// getStaticPodTerminationGracePeriodSecondsForRevision reads the static pod manifest from a configmap
+// in the target namespace and returns the value of terminationGracePeriodSeconds.
+// In case no value was provided for the static pod it returns a default value of 30 seconds.
+func (c *missingStaticPodController) getStaticPodTerminationGracePeriodSecondsForRevision(revision int) (time.Duration, error) {
```

Comment:
can we add a TODO with David's suggestion for the future? https://github.com/openshift/library-go/pull/1316/files#r805889436